### ISSUE TITLE
Bite the bullet and upgrade Python?

### ISSUE THREAD

**Issue by edeutsch** [RTXteam/RTX#1889]
I have been working with Richard to upgrade our reasoner-validator. ARAX validates all TRAPI as part of the ARS support system. Richard has been working on new features to do semantic validation, not just JSON schema level validation.

With the import of Biolink model toolkits, the dependencies can no longer be resolved under Python 3.7. It seems that Python 3.8 at least is required to support the new validator.

Can we consider upgrading our system from Python 3.7.12 to 3.8 or greater? I vaguely think that the ITRB instances are already running 3.8? Does anyone know which versions they're running?

Any comments or recommendations or insights?

**Comment by edeutsch** [comment]
Does anyone have tales to tell about successfully routinely running ARAX on their own machines under Python 3.x.y?

**Comment by amykglen** [comment]
I routinely run Python 3.9.7, no issues. `pip install -r requirements.txt` has always had issues for me relating to some obo toolkit thing (I think not just for Python 3.9? not sure), but it's never affected running ARAX.

**Comment by edeutsch** [comment]
I propose we upgrade the custom Python version on arax.ncats.io to Python 3.8.13, which is the latest 3.8.x. Although I'm not certain, it seems like 3.8 is a very common baseline and I think that's what our ITRB instances are running? And I think that will resolve the current problem.

What do you think?

Or should we be more ambitious?

**Comment by edeutsch** [comment]
hmm, after seeing that it seems that the Biolink folks seem to be moving their stuff to Python 3.9, I'm not thinking we may as well move to 3.9 rather than moving to something that is pretty long in the tooth already (3.8)

**Comment by amykglen** [comment]
I recall last time that we thought ITRB is running Python 3.8, but now I can't find where we saw that... our Dockerfiles seem to specify Python 3.7

**Comment by edeutsch** [comment]
So I have built Python 3.9.13 on arax.ncats.io in
/mnt/data/python/Python-3.9.13

/test, /beta, /kg2beta, /devED, /devLM all seem to be running fine on this version of Python.

One thing that is majorly different is that I ignored all the version information in requirements.txt and just installed the latest of everything. And the basic things seems to work.

So my main questions are:
1) Okay to continue rolling out 3.9.13 everywhere, or are there concerns?
2) Okay to just upgrade everything to the latest packages and re-freeze that state to requirements.txt? Or how do we want to manage package versions in requirements.txt more carefully? I think many of those package versions in our requirements.txt are pretty ancient.

Tests all pass, but seem glacial.

What do you think?

**Comment by edeutsch** [comment]
Just to keep everyone in the loop, the requirements.txt has been majorly updated. I installed the latest Python 3.9.13 on arax.ncats.io and installed everything in requirements.txt without specifying a version number so that we get the latest and then updated our current requirements.txt to reflect that freeze state: https://github.com/RTXteam/RTX/commit/f87a534229744630263f2edab45f3a136bba51f4. This has the side effect of making dependabot happy with all 4 things that it was all huffy about. all seems to work. with Amy's help, we have started ITRB deployments using Python 3.9. arax.ci.transltr.io is already working fine. The others are not deployed yet, pending some other fixes, but should be fine when we get to it.
